src/canister_id.rs

```diff
@@ -0,0 +1,246 @@
+use hyper::{header::HOST, Request, Uri};
+use ic_agent::export::Principal;
+
+use crate::config::dns_canister_config::DnsCanisterConfig;
+
+/// A resolver for `Principal`s from a `Uri`.
+trait UriCanisterIdResolver: Sync + Send {
+    fn resolve(&self, uri: &Uri) -> Option<Principal>;
+}
+
+impl<T: UriCanisterIdResolver> UriCanisterIdResolver for &T {
+    fn resolve(&self, uri: &Uri) -> Option<Principal> {
+        T::resolve(self, uri)
+    }
+}
+
+struct UriParameterResolver;
+
+impl UriCanisterIdResolver for UriParameterResolver {
+    fn resolve(&self, uri: &Uri) -> Option<Principal> {
+        url::form_urlencoded::parse(uri.query()?.as_bytes())
+            .find(|(name, _)| name == "canisterId")
+            .and_then(|(_, canister_id)| Principal::from_text(canister_id.as_ref()).ok())
+    }
+}
+
+impl UriCanisterIdResolver for DnsCanisterConfig {
+    fn resolve(&self, uri: &Uri) -> Option<Principal> {
+        self.resolve_canister_id(uri.host()?)
+    }
+}
+
+/// A resolver for `Principal`s from a `Request`.
+pub trait CanisterIdResolver<B>: Sync + Send {
+    fn resolve(&self, request: &Request<B>) -> Option<Principal>;
+}
+
+impl<B, T: CanisterIdResolver<B>> CanisterIdResolver<B> for &T {
+    fn resolve(&self, request: &Request<B>) -> Option<Principal> {
+        T::resolve(self, request)
+    }
+}
+
+struct RequestUriResolver<T>(pub T);
+
+impl<B, T: UriCanisterIdResolver> CanisterIdResolver<B> for RequestUriResolver<T> {
+    fn resolve(&self, request: &Request<B>) -> Option<Principal> {
+        self.0.resolve(request.uri())
```

Comment:
Does it make sense for this to just be `UriResolver` and then you pass it a `Uri` instead of a full on `Request`? I just notice that we _only_ use the request for calling `.uri`.

Reply:
No, that's what `UriCanisterIdResolver` is for. This class is the bridge between those two.

Reply:
Ah, my bad! But then can't you just call `UriCanisterIdResolver` with the `request.uri()` instead?